File: src/main/kotlin/graph/서로소_집합_알고리즘.java
package graph;

import java.util.*;

public class 서로소_집합_알고리즘 {

    private static int[] parent;

    public static int findParent(int x){
        if(parent[x] == x) return x;
        return parent[x] = findParent(parent[x]);
    }

    public static void unionParent(int a, int b){
        a = findParent(a);
        b = findParent(b);
        if(a < b) parent[b] = a;
        else parent[a] = b;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();   //노드의 개수
        int m = sc.nextInt();   //간선의 개수

        //부모 테이블 자기 자신으로 초기화
        parent = new int[n+1];
        for(int i = 1; i <= n; i++){
            parent[i] = i;
        }

        //간선정보 입력받기
        for(int i = 0; i < m; i++){
            int a = sc.nextInt();
            int b = sc.nextInt();

            unionParent(a, b);
        }

        //각 원소가 속한 집합 출력
        for(int i = 1; i <= n; i++){
            System.out.print(findParent(i) + " ");
        }
        System.out.println();

        //부모 테이블 내용 출력
        for(int i = 1; i <= n; i++){
            System.out.print(parent[i] + " ");
        }
    }
}

File: src/main/kotlin/graph/서로소_집합_알고리즘.kt
package graph

import java.util.*
import kotlin.math.*

fun main(){
    val sc = Scanner(System.`in`)
    val v = sc.nextInt()    //노드의 개수
    val e = sc.nextInt()    //간선의 개수

    val parent = IntArray(v + 1)
    //부몰 테이블에서 부모를 자기 자신으로 초기화
    for(i in 1..v){
        parent[i] = i
    }

    fun findParent(x: Int): Int {
        //루트 노드가 아니라면 루트 노드를 찾을 때까지 재귀적으로 호출
        if(x == parent[x]) return x
        return findParent(parent[x]).also { parent[x] = it }
    }

    fun unionParent(a: Int, b: Int){
        val parentA = findParent(a)
        val parentB = findParent(b)
        //부모의 인덱스가 작은 것으로 합침
        if(parentA < parentB) parent[parentB] = parent[parentA] else parent[parentA] = parent[parentB]
    }

    //Union 연산을 각각 수행
    for(i in 0 until e){
        val a = sc.nextInt()
        val b = sc.nextInt()
        unionParent(a, b)
    }

    //각 원소가 속한 집합 출력
    for(i in 1..v){
        print("${findParent(i)} ")
    }
    println()

    //부모 테이블 내용 출력
    for(i in 1..v){
        print("${parent[i]} ")
    }
    println()
}